### PR TITLE
Fix #2133: apply_patch fails

### LINKED ISSUE #2133: apply_patch fails
### What happened?

gpt-5.5 seems to have issues with apply_patch

I’m considering adding request_id: 1 and calling mark before each test. This seems necessary because there are many tests involved. However, I want to ensure this doesn’t complicate things too much. I’ll need to think through how many tests this change affects and if there are any implications for overall functionality. It's important to maintain clarity and avoid introducing any unexpected issues from this adjustment. Let's see how this goes! **Updating processes**

I’m looking to compile information first and ensure that I update the messages conversion properly. There’s definitely some work to be done to patch up operations. It’s important I get this right so everything flows smoothly going forward. I wonder if I’m missing any details. I’ll need to stay organized and maybe make some notes to keep track of the steps I take. Let’s move forward with this!

 Evaluating request_id logic: I think it's time for me to update helpers and tests to properly mark certain situations. Currently, many tests don't do that, so it’s a bit broad in scope. For instance, I’m considering having `IssueListLoaded` apply only if the request_id is 0 and there are no pending items, but that introduces production risks. Alternately, should I define this in the reducer with a fallback for the current repo or filter? Hmm, it makes me rethink if request_id starting at 1 might be a more robust solution.

 Evaluating request_id logic: I think it's time for me to update helpers and tests to properly mark certain situations. Currently, many tests don't do that, so it’s a bit broad in scope. For instance, I’m considering having `IssueListLoaded` apply only if the request_id is 0 and there are no pending items, but that introduces production risks. Alternately, should I define this in the reducer with a fallback for the current repo or filter? Hmm, it makes me rethink if request_id starting at 1 might be a more robust solution. **Refining request handling**

I’m contemplating using “pending” if it's present, but should I also accept the same scope or filter? If production always marks pending, that could reduce issues, although stale responses might still apply if pending items are cleared. I’m considering a helper function for tests, which might be manageable with a regex to insert request_id: 0. However, it gets tricky since each test uses IssueListLoaded directly. I might think about setting `next_issue_list_request_id` to 1 in the default IssueState, but I need to prioritize patching production first and checking for any errors.

 Inspecting test failures: I need to dig deeper into why the tests are failing, even though the compile check passed. It might be because the cargo check didn’t compile the tests properly. I’ll inspect the types to see if perhaps the recent patch didn’t successfully add the request_id. Maybe the earlier patch didn't stick as expected? I should review the types around the event to get a clearer understanding of the issue.

 Troubleshooting patch issues: I need to keep moving forward. I’m trying to figure out why the types patch didn't stick. The read shows no request_id in the enum, but the earlier apply said it was successful. Is it possible that the patch was applied to another area? I'm going to do a search to clarify this confusion and see what I can find. I really want to get to the bottom of this!

 Inspecting apply_patch output: I’m looking into a weird situation where the apply_patch output says it succeeded, but the changes aren't in the file. Hmm, could it be stale? I need to edit it carefully to avoid mistakes. Also, it seems that list_reload_pending might not be in the types I expect. I should inspect the surrounding IssuesState to see if there's something off there. Let's figure this out step by step!

 Addressing patch issues: I need to continue from where I left off and really focus on editing the types. It seems like the patch I applied earlier didn’t stick, which is kind of strange — maybe it failed without any notification. It says there’s no request_id, so I need to be careful with the replace/ast process. I also need to update the messages and check if anything is stuck to ensure smooth progress.

 Updating tests and messages: I'm considering how to approach an issue with an explicit request for identity. It's probably better not to go down that route. If the construction in state tests uses 0, it could create a backdoor where any event with 0 applies. That might complicate messages in production. It's possible to convert messages into 0, but maybe the code could dispatch them. I think it’s best to update the tests first, specifically fixing types and messages with Python.

 Fixing errors in tools: I need to continue working on this. It sounds like I need to use the tools available, but the current state is messy with many errors. It looks like previous patches weren't fully applied to the issues in ops and messages conversion, possibly due to something being overwritten. I’ll need to fix these issues carefully and make sure to inspect the relevant files to get everything back on track!

DETERMINE IF THERE IS A TOOL ISSUE

### What did you expect to happen?

apply it and not get messed up

### Client information

Image

### Login information

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Fixes #2133

This PR hardens apply_patch so it no longer reports success after silently applying only the first file section from a multi-file unified diff. It keeps the tool aligned with its single-target file schema by rejecting multi-file patches, validating patch headers against the requested target path, and returning a dedicated patch-apply failure error when diff context cannot be applied.

## Dive Deeper

Issue #2133 pointed to confusing apply_patch behavior where a successful result could be returned even though expected changes were missing. The underlying problem was that Diff.parsePatch can return multiple structured patches, but the implementation only used the first parsed patch. A patch containing multiple file sections could therefore partially apply and still look successful.

The implementation now:

- Rejects empty or unparseable patch content with INVALID_TOOL_PARAMS.
- Rejects multi-file unified diffs and instructs callers to use one apply_patch call per target file.
- Validates that the parsed patch header matches the provided absolute_path or file_path target before applying changes.
- Preserves valid single-file behavior, including multi-hunk patches and new-file patches from /dev/null.
- Adds PATCH_APPLY_FAILURE so context-mismatch failures are no longer reported as EDIT_NO_CHANGE.
- Tightens constructor duck-typing so partial service-shaped objects are not misclassified.
- Waits for LSP diagnostics using the actual file path written by the tool instead of resolving paths from patch headers.

Focused regression coverage was added for the issue scenarios and related edge cases in packages/tools/src/__tests__/apply-patch.test.ts.

## Reviewer Test Plan

Recommended validation:

- Run the focused apply_patch tests:

      npm run test --workspace @vybestack/llxprt-code-tools -- apply-patch

- Run the tools package verification commands:

      npm run test --workspace @vybestack/llxprt-code-tools
      npm run lint --workspace @vybestack/llxprt-code-tools
      npm run typecheck --workspace @vybestack/llxprt-code-tools

Manual review focus:

- Confirm multi-file patches are rejected before any partial write occurs.
- Confirm valid single-file and multi-hunk patches still apply successfully.
- Confirm patch header mismatches leave the target file unchanged.
- Confirm context mismatches return PATCH_APPLY_FAILURE.

## Testing Matrix

|          | APPLE | WIN | LINUX |
| -------- | ----- | --- | ----- |
| npm run  | ?     | ?   | ?     |
| npx      | ?     | ?   | ?     |
| Docker   | ?     | ?   | ?     |
| Podman   | ?     | -   | -     |
| Seatbelt | ?     | -   | -     |

## Linked issues / bugs

Fixes #2133

### CODE REVIEW

File: packages/tools/src/__tests__/apply-patch.test.ts [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+      writeFileSync(target, 'unchanged\n', 'utf-8');
+
+      const result = await executePatch({
+        absolute_path: target,
+        patch_content: 'this is not a patch at all',
+      });
+
+      expect(result.error).toBeDefined();
+      expect(result.error?.type).toBe(ToolErrorType.INVALID_TOOL_PARAMS);
+    });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/__tests__/apply-patch.test.ts:319]: Addressed. Both assertions are now locked in:

- The empty-patch case asserts `ToolErrorType.INVALID_TOOL_PARAMS` (in addition to the message and file-unchanged checks).
- The malformed-patch case asserts the target file is left unchanged.

To make the empty-patch type assertion actually hold, the test helper now routes through `validateBuildAndExecute` (the production call path) rather than a `try/catch` around `build().execute()`. The previous helper caught the validation `throw` from `build()` and returned an error with no `type`, so the assertion would not have passed against the real tool. `validateBuildAndExecute` surfaces validation/execution errors with their real `ToolErrorType`, matching runtime behavior. Full test/lint/typecheck/build suite passes.

File: packages/tools/src/tools/apply-patch.ts [hunk 2]
```diff
@@ -535,12 +574,87 @@ class ApplyPatchToolInvocation extends BaseToolInvocation<
...
+      // directory's same-named file. Only headers without a directory component
+      // fall back to basename comparison.
+      if (header.includes('/') || header.includes(path.sep)) {
+        return header === relativePath;
+      }
+      return path.basename(header) === targetName;
+    };
+
+    const newMatches = headerMatches(newHeader);
+    const oldMatches = headerMatches(oldHeader);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/tools/src/tools/apply-patch.ts]: Addressed. Both sides are now normalized to forward slashes before comparison via a `toPosix` helper (`p.split(path.sep).join('/')`). On Windows this converts the backslashes returned by `path.relative` so that `header === relativePath` matches against the always-forward-slash diff headers; on POSIX it is a no-op (`path.sep === '/'`), so existing behavior is unchanged. The directory-qualified header match/reject tests continue to pass.
